File: Java/beadsort/AmazonQ/AmazonQ.java
import java.util.Random;

public class AmazonQ {
    public static void beadSort(int[] arr) {
        if (arr.length == 0) return;

        // Find the maximum element
        int max = arr[0];
        for (int i = 1; i < arr.length; i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }

        // Create a grid for beads
        boolean[][] grid = new boolean[arr.length][max];

        // Drop the beads
        for (int i = 0; i < arr.length; i++) {
            for (int j = 0; j < arr[i]; j++) {
                grid[i][j] = true;
            }
        }

        // Let beads fall to the bottom
        for (int j = 0; j < max; j++) {
            int count = 0;
            for (int i = 0; i < arr.length; i++) {
                if (grid[i][j]) {
                    count++;
                }
            }
            // Count beads in each column
            for (int i = arr.length - 1; i >= arr.length - count; i--) {
                grid[i][j] = true;
            }
            for (int i = arr.length - count - 1; i >= 0; i--) {
                grid[i][j] = false;
            }
        }

        // Put the sorted numbers back into array
        for (int i = 0; i < arr.length; i++) {
            int count = 0;
            for (int j = 0; j < max; j++) {
                if (grid[i][j]) {
                    count++;
                }
            }
            arr[i] = count;
        }
    }

    public static void main(String[] args) {
        int[] numbers = new int[1000];
        Random random = new Random();
        
        // Generate 1000 random numbers between 1 and 1000
        for (int i = 0; i < numbers.length; i++) {
            numbers[i] = random.nextInt(1000) + 1;
        }

        // Sort the array using bead sort
        beadSort(numbers);
    }
}

File: C/beadsort/amazonQ/amazonq.c
#include <stdlib.h>
#include <time.h>

void beadsort(int *arr, int len) {
    int max = arr[0];
    
    // Find maximum element
    for (int i = 1; i < len; i++) {
        if (arr[i] > max) {
            max = arr[i];
        }
    }
    
    // Allocate memory for beads
    unsigned char *beads = (unsigned char *)calloc(max * len, sizeof(unsigned char));
    if (beads == NULL) return;  // Memory allocation check
    
    // Set beads for each number
    for (int i = 0; i < len; i++) {
        for (int j = 0; j < arr[i]; j++) {
            beads[i * max + j] = 1;
        }
    }
    
    // Let beads fall
    for (int j = 0; j < max; j++) {
        int sum = 0;
        for (int i = 0; i < len; i++) {
            sum += beads[i * max + j];
            beads[i * max + j] = 0;
        }
        
        // Count beads after falling
        for (int i = len - 1; i >= len - sum; i--) {
            beads[i * max + j] = 1;
        }
    }
    
    // Put sorted values back in array
    for (int i = 0; i < len; i++) {
        int count = 0;
        for (int j = 0; j < max; j++) {
            count += beads[i * max + j];
        }
        arr[i] = count;
    }
    
    free(beads);
}

int main() {
    const int SIZE = 1000;
    int arr[SIZE];
    
    // Initialize random seed
    srand((unsigned int)time(NULL));
    
    // Generate 1000 random numbers
    for (int i = 0; i < SIZE; i++) {
        arr[i] = rand() % 1000;  // Random numbers between 0 and 999
    }
    
    // Sort the array
    beadsort(arr, SIZE);
    
    return 0;
}
